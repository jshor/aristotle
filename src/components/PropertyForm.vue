<template>
  <div class="property-form">
    <div
      v-for="(data, propertyName) in model"
      :key="propertyName"
      class="property-form__entry"
    >
      <label
        :for="`${id}_${propertyName}`"
        :class="{
          'property-form__label--disabled': data.disabled
        }"
        class="property-form__label"
      >
        {{ data.label }}
      </label>

      <!-- drop-down select -->
      <select
        v-if="data.options"
        v-model="data.value"
        :id="`${id}_${propertyName}`"
        :disabled="data.disabled"
        class="property-form__input"
        @change="onChange"
      >
        <option
          v-for="(value, key) in data.options"
          :key="key"
          :value="value"
        >
          {{ key }}
        </option>
      </select>

      <!-- toggle switch -->
      <div
        v-else-if="data.type === 'boolean'"
        class="property-form__switch"
      >
        <property-switch
          v-model.boolean="data.value"
          :id="`${id}_${propertyName}`"
          :disabled="data.disabled"
          @update:modelValue="onSwitch(propertyName)"
        />
      </div>

      <!-- text boxes -->
      <input
        v-else
        v-model="data.value"
        :min="data.min"
        :max="data.max"
        :type="data.type"
        :id="`${id}_${propertyName}`"
        :disabled="data.disabled"
        class="property-form__input"
        @change="onChange"
      />

      <small
        v-if="data.description"
        class="property-form__description"
      >
        {{ data.description }}
      </small>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType, ref, watch } from 'vue'
import cloneDeep from 'lodash.clonedeep'
import PropertySwitch from '@/components/properties/PropertySwitch.vue'
import Icon from '@/components/Icon.vue'
import isMobile from '@/utils/isMobile'

export default defineComponent({
  name: 'PropertyForm',
  components: {
    Icon,
    PropertySwitch
  },
  props: {
    modelValue: {
      type: Object as PropType<PropertySet>,
      default: () => ({})
    },
    id: {
      type: String,
      default: ''
    }
  },
  setup (props, { emit }) {
    const getModel = (model: PropertySet): PropertySet => Object
      .keys(model)
      // filter to ensure mobile-only or desktop-only settings are visible appropriately
      .filter(key => isMobile
        ? !model[key].desktopOnly
        : !model[key].mobileOnly)
      .reduce((m, k) => ({
        ...m,
        [k]: cloneDeep(model[k])
      }), {})

    const model = ref<PropertySet>(getModel(props.modelValue))

    watch(() => props.modelValue, value => {
      model.value = getModel(value)
    }, { deep: true })

    function onSwitch (propertyName: string) {
      const property = model.value[propertyName]

      property.excludes?.forEach(p => {
        if (property.value && model.value[p].type === 'boolean') {
          model.value[p].value = false
        }
      })

      onChange()
    }

    function onChange () {
      emit('update:modelValue', model.value)
    }

    return {
      model,
      close,
      onSwitch,
      onChange
    }
  }
})
</script>

<style lang="scss">
.property-form {
  padding: 0.25em;
  color: var(--color-primary);

  &__icon {
    cursor: pointer;
    transition: all 0.25s;

    &:hover, &:focus {
      transform: scale(2);
    }
  }

  &__entry {
    display: grid;
    grid-template-columns: 50% 50% 100%;
  }

  &__label--disabled {
    color: var(--color-secondary);
  }

  &__input, &__label, &__switch, &__description {
    max-width: 100%;
    margin: 0.25em;
  }

  &__input {
    box-sizing: border-box;
    background-color: var(--color-bg-primary);
    border: 1px solid var(--color-primary);
    color: var(--color-primary);
    font-family: system-ui, sans-serif;
    padding: 0.25em;
      width: auto;
      height: auto;

    &[type="color"] {
      padding: 0;

      &::-webkit-color-swatch-wrapper {
        padding: 0;
      }
      &::-webkit-color-swatch {
        border: none;
      }
    }
  }

  &__switch {
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }

  &__description {
    width: 100%;
    grid-column: 1/3;
    color: var(--color-secondary);
  }
}
</style>
