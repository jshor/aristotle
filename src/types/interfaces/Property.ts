type PropertyBase<T> = {
  label: string
  description?: string
  options?: {
    [text: string]: T
  }
  mobileOnly?: boolean
  desktopOnly?: boolean
  disabled?: boolean
  excludes?: string[]
}

/** A numeric property. */
type NumericProperty = {
  /** A nnnnn or a color. */
  type: 'number'
  value: number
  min?: number
  max?: number
} & PropertyBase<number>

/** A string property. */
type StringProperty = {
  /** A text or a color. */
  type: 'text' | 'color'
  value: string
} & PropertyBase<string>

/** A boolean property. */
type BooleanProperty = {
  type: 'boolean'
  value: boolean
} & PropertyBase<boolean>

type Property = NumericProperty | StringProperty | BooleanProperty

export default Property
