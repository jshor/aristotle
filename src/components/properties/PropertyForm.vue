<template>
  <div class="property-form">
    <div
      v-for="(data, propertyName) in (model as Required<PropertyList>)"
      :key="propertyName"
      class="property-form__entry"
    >
      <label
        :for="`${id}_${propertyName}`"
        :class="{
          'property-form__label--disabled': data.disabled
        }"
        class="property-form__label"
        data-test="label"
      >
        {{ data.label }}
      </label>

      <!-- drop-down select -->
      <select
        v-if="data.options"
        v-model="data.value"
        :id="`${id}_${propertyName}`"
        :disabled="data.disabled"
        class="property-form__input"
        @change="onChange"
      >
        <option
          v-for="(value, key) in data.options"
          :key="key"
          :value="value"
        >
          {{ key }}
        </option>
      </select>

      <!-- toggle switch -->
      <div
        v-else-if="data.type === 'boolean'"
        class="property-form__switch"
      >
        <property-switch
          v-model.boolean="(data.value as boolean)"
          :id="`${id}_${propertyName}`"
          :disabled="data.disabled"
          @update:modelValue="onSwitch(propertyName)"
        />
      </div>

      <!-- numeric input -->
      <input
        v-else-if="data.type === 'number'"
        v-model="data.value"
        :min="data.min"
        :max="data.max"
        type="number"
        :id="`${id}_${propertyName}`"
        :disabled="data.disabled"
        class="property-form__input"
        @change="onChange"
      />

      <!-- text boxes -->
      <input
        v-else
        v-model="data.value"
        :type="data.type"
        :id="`${id}_${propertyName}`"
        :disabled="data.disabled"
        class="property-form__input"
        @change="onChange"
      />

      <small
        v-if="data.description"
        class="property-form__description"
      >
        {{ data.description }}
      </small>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, watch } from 'vue'
import cloneDeep from 'lodash.clonedeep'
import PropertySwitch from '@/components/properties/PropertySwitch.vue'
import isMobile from '@/utils/isMobile'
import Property from '@/types/interfaces/Property'

type PropertyList = Record<string, Property<string | boolean | number>>

const props = withDefaults(defineProps<{
  /** Set of properties. */
  modelValue: PropertyList
  /** Custom ID for this property form. */
  id: string
}>(), {
  modelValue: () => ({}),
  id: ''
})

const emit = defineEmits<{
  /** Model update. */
  (e: 'update:modelValue', properties: PropertyList): void
}>()

const model = ref<PropertyList>(getModel(props.modelValue))

watch(() => props.modelValue, value => {
  model.value = getModel(value)
}, { deep: true })

function onSwitch (propertyName: keyof PropertyList) {
  const property = model.value[propertyName]

  property?.excludes?.forEach((p: keyof PropertyList) => {
    if (property.value && model.value[p].type === 'boolean') {
      model.value[p].value = false
    }
  })

  onChange()
}

function getModel (model: PropertyList) {
  const properties: PropertyList = {}

  for (const key in model) {
    const k = key as keyof PropertyList
    const isVisible = isMobile
      ? !model[k]!.desktopOnly
      : !model[k]!.mobileOnly

    if (isVisible) {
      properties[k] = cloneDeep(model[k])
    }
  }

  return properties
}

function onChange () {
  validate()
  emit('update:modelValue', model.value)
}

function validate () {
  for (const key in model.value) {
    const k = key as keyof PropertyList

    if (!isValidValue(model.value[k])) {
      model.value[k].value = props.modelValue[k].value
    }
  }
}

function isValidValue (property: Property<string | boolean | number>) {
  if (property.type === 'boolean') return true
  if (property.type === 'number') {
    const value = property.value as number

    if (property.max !== undefined && property.max < value) return false
    if (property.min !== undefined && property.min > value) return false

    return true
  }

  return !!property.value
}
</script>

<style lang="scss">
.property-form {
  padding: 0.25em;
  padding-top: 0;
  color: var(--color-primary);

  &__icon {
    cursor: pointer;
    transition: all 0.25s;

    &:hover, &:focus {
      transform: scale(2);
    }
  }

  &__entry {
    display: grid;
    grid-template-columns: 50% 50% 100%;
    margin-bottom: 0.5em;
  }

  &__label--disabled {
    color: var(--color-secondary);
  }

  &__input, &__label, &__switch, &__description {
    max-width: 100%;
    margin: 0.25em;
  }

  &__input {
    box-sizing: border-box;
    background-color: var(--color-bg-primary);
    border: 1px solid var(--color-primary);
    color: var(--color-primary);
    font-family: system-ui, sans-serif;
    padding: 0.25em;
    width: auto;
    height: auto;

    &[type="color"] {
      padding: 0;

      &::-webkit-color-swatch-wrapper {
        padding: 0;
      }
      &::-webkit-color-swatch {
        border: none;
      }
    }
  }

  &__switch {
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }

  &__description {
    width: 100%;
    grid-column: 1/3;
    color: var(--color-secondary);
  }
}
</style>
