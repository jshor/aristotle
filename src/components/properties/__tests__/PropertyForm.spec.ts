import { mount, VueWrapper } from '@vue/test-utils'
import PropertyForm from '../PropertyForm.vue'
import Property from '@/types/interfaces/Property'

describe('Properties Form', () => {
  let wrapper: VueWrapper
  let model: Record<string, Property<string | boolean | number>>

  beforeEach(() => {
    model = {
      name: {
        label: 'Name',
        type: 'text',
        value: 'My Property',
        description: 'A basic text input.'
      },
      count: {
        label: 'Count',
        type: 'number',
        value: 10,
        description: 'A basic numeric input.',
        min: 1,
        max: 50
      },
      switch1: {
        label: 'Switch 1',
        type: 'boolean',
        value: false,
        description: 'A basic switch.',
        excludes: ['switch2']
      },
      switch2: {
        label: 'Switch 2',
        type: 'boolean',
        value: true
      },
      options: {
        label: 'Select an option',
        type: 'text',
        value: 'option1',
        description: 'An option applies.',
        options: {
          'Option 1': 'option1',
          'Option 2': 'option2'
        }
      }
    }

    wrapper = mount(PropertyForm, {
      props: {
        modelValue: model
      }
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render when no properties are provided', async () => {
    await wrapper.setProps({
      modelValue: undefined
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should prefix property IDs with the prop ID', async () => {
    await wrapper.setProps({
      modelValue: model,
      id: 'prop'
    })

    expect(wrapper.find('input').attributes('id')).toBe('prop_name')
  })

  it('should show the label as disabled when the property is disabled', async () => {
    await wrapper.setProps({
      modelValue: Object
        .keys(model)
        .reduce((m, key) => ({
          ...m,
          [key]: {
            ...model[key],
            disabled: true
          }
        }), {})
    })

    wrapper.findAll('[data-test="label"]').forEach(label => {
      expect(label.classes()).toContain('property-form__label--disabled')
    })
  })

  describe('when the text input is changed', () => {
    it('should update the model with the valid input', async () => {
      await wrapper
        .find('input[type="text"]')
        .setValue('some value')

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        name: expect.objectContaining({
          value: 'some value'
        })
      })])
    })

    it('should reset the value of the text when it is entered as blank', async () => {
      await wrapper
        .find('input[type="text"]')
        .setValue('')

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        name: expect.objectContaining({
          value: 'My Property'
        })
      })])
    })

    it('should reset the value of a number that exceeds the max', async () => {
      await wrapper
        .find('input[type="number"]')
        .setValue(60)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        count: expect.objectContaining({
          value: 10
        })
      })])
    })
  })

  describe('when the number input is updated', () => {
    it('should update the valid value of the input number', async () => {
      await wrapper
        .find('input[type="number"]')
        .setValue(20)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        count: expect.objectContaining({
          value: 20
        })
      })])
    })

    it('should reset the value of a number that is less than the min', async () => {
      await wrapper
        .find('input[type="number"]')
        .setValue(-20)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        count: expect.objectContaining({
          value: 10
        })
      })])
    })

    it('should reset the value of a number that exceeds the max', async () => {
      await wrapper
        .find('input[type="number"]')
        .setValue(60)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        count: expect.objectContaining({
          value: 10
        })
      })])
    })
  })

  describe('when a switch is toggled', () => {
    it('should update the value', async () => {
      await wrapper
        .findComponent({ name: 'PropertySwitch' })
        .vm
        .$emit('update:modelValue', true)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        switch1: expect.objectContaining({
          value: true
        })
      })])
    })

    it('should turn off an excluded switch', async () => {
      await wrapper
        .findComponent({ name: 'PropertySwitch' })
        .vm
        .$emit('update:modelValue', true)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        switch1: expect.objectContaining({
          value: true
        }),
        switch2: expect.objectContaining({
          value: false
        })
      })])
    })

    it('should never turn on an excluded switch', async () => {
      await wrapper.setProps({
        switch1: {
          ...model.switch1,
          value: true
        }
      })
      await wrapper
        .findComponent({ name: 'PropertySwitch' })
        .vm
        .$emit('update:modelValue', true)

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        switch2: expect.objectContaining({
          value: false
        })
      })])
    })
  })

  describe('when a selection option is changed', () => {
    it('should update the value', async () => {
      await wrapper
        .find('select')
        .setValue('option2')

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual([expect.objectContaining({
        options: expect.objectContaining({
          value: 'option2'
        })
      })])
    })
  })

  describe('when in desktop mode', () => {
    it('should not show a property that is mobile-only', async () => {
      await wrapper.setProps({
        modelValue: {
          name: {
            ...model.name,
            mobileOnly: true
          }
        }
      })

      expect(wrapper.find('input').exists()).toBe(false)
    })
  })
})
