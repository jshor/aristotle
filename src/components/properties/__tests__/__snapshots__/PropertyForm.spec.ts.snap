// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Properties Form > should match the snapshot 1`] = `
<div
  class="property-form"
>
  
  <div
    class="property-form__entry"
  >
    <label
      class="property-form__label"
      data-test="label"
      for="_name"
    >
      Name
    </label>
    <!-- drop-down select -->
    
    <!-- text boxes -->
    <input
      class="property-form__input"
      id="_name"
      type="text"
    />
    
    <small
      class="property-form__description"
    >
      A basic text input.
    </small>
  </div>
  <div
    class="property-form__entry"
  >
    <label
      class="property-form__label"
      data-test="label"
      for="_count"
    >
      Count
    </label>
    <!-- drop-down select -->
    
    <!-- numeric input -->
    <input
      class="property-form__input"
      id="_count"
      max="50"
      min="1"
      type="number"
    />
    
    <small
      class="property-form__description"
    >
      A basic numeric input.
    </small>
  </div>
  <div
    class="property-form__entry"
  >
    <label
      class="property-form__label"
      data-test="label"
      for="_switch1"
    >
      Switch 1
    </label>
    <!-- drop-down select -->
    
    <!-- toggle switch -->
    <div
      class="property-form__switch"
    >
      <label
        class="property-switch"
        modelmodifiers="[object Object]"
        tabindex="0"
      >
        <input
          class="property-switch__input"
          id="_switch1"
          type="checkbox"
        />
        <div
          class="property-switch__trigger"
        />
      </label>
    </div>
    
    <small
      class="property-form__description"
    >
      A basic switch.
    </small>
  </div>
  <div
    class="property-form__entry"
  >
    <label
      class="property-form__label"
      data-test="label"
      for="_switch2"
    >
      Switch 2
    </label>
    <!-- drop-down select -->
    
    <!-- toggle switch -->
    <div
      class="property-form__switch"
    >
      <label
        class="property-switch property-switch--on"
        modelmodifiers="[object Object]"
        tabindex="0"
      >
        <input
          class="property-switch__input"
          id="_switch2"
          type="checkbox"
        />
        <div
          class="property-switch__trigger property-switch__trigger--on"
        />
      </label>
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="property-form__entry"
  >
    <label
      class="property-form__label"
      data-test="label"
      for="_options"
    >
      Select an option
    </label>
    <!-- drop-down select -->
    <select
      class="property-form__input"
      id="_options"
    >
      
      <option
        value="option1"
      >
        Option 1
      </option>
      <option
        value="option2"
      >
        Option 2
      </option>
      
    </select>
    <small
      class="property-form__description"
    >
      An option applies.
    </small>
  </div>
  
</div>
`;

exports[`Properties Form > should render when no properties are provided 1`] = `"<div class="property-form"></div>"`;
